socorro/processor/mozilla_transform_rules.py

```diff
@@ -816,3 +816,79 @@ def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+
+
+#==============================================================================
+class BetaVersionRule(Rule):
+    required_config = Namespace()
+    required_config.add_option(
+        'database_class',
+        doc="the class of the database",
+        default=
+        'socorro.external.postgresql.connection_context.ConnectionContext',
+        from_string_converter=str_to_python_object,
+        reference_value_from='resource.postgresql',
+    )
+    required_config.add_option(
+        'transaction_executor_class',
+        default="socorro.database.transaction_executor."
+                "TransactionExecutorWithInfiniteBackoff",
+        doc='a class that will manage transactions',
+        from_string_converter=str_to_python_object,
+        reference_value_from='resource.postgresql',
+    )
+
+    #--------------------------------------------------------------------------
+    def __init__(self, config):
+        super(BetaVersionRule, self).__init__(config)
+        database = config.database_class(config)
+        transaction = config.transaction_executor_class(
+            config,
+            database,
+        )
+        sql = """
+            SELECT
+                pv.product_name,
+                pv.release_version,
+                pv.build_type,
+                pvb.build_id,
+                pv.version_string
+            FROM product_versions pv
+                LEFT JOIN product_version_builds pvb ON
+                    (pv.product_version_id = pvb.product_version_id)
+        """
+        results = transaction(
+            execute_query_fetchall,
+            sql
+        )
+        self.versions_data = {}
+        for product, version, channel, build, real_version in results:
+            key = '%s:%s:%s:%s' % (product, version, channel, build)
+            self.versions_data[key] = real_version
+
+    #--------------------------------------------------------------------------
+    def version(self):
+        return '1.0'
+
+    #--------------------------------------------------------------------------
+    def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+        try:
+            key = '%s:%s:%s:%s' % (
+                processed_crash['product'],
+                processed_crash['version'],
+                processed_crash['release_channel'],
+                processed_crash['build'],
+            )
+            real_version = self.versions_data.get(key)
+
+            if real_version:
+                processed_crash['version'] = real_version
+            elif processed_crash['release_channel'].lower() == 'beta':
+                # This is a beta version but we do not have data about it. It
+                # could be because we don't have it yet (if the cron jobs are
+                # running late for example), so we mark this crash. This way,
+                # we can reprocess it later to give it the correct version.
+                processed_crash['version'] += 'b0'
```

Comment:
if this crash is reprocessed for some reason other than this _unknown beta_ situation and it is still an unknown beta, then this line will tag an additional 'b0' to the end of the version.